File: src/main/java/sample/engine/canvasui/painter/Painter.java
package sample.engine.canvasui.painter;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;
import javafx.scene.paint.Paint;
import javafx.scene.text.TextAlignment;
import sample.engine.canvasui.painter.actions.*;
import sample.engine.images.Rect;

import java.util.ArrayList;
import java.util.List;

public class Painter {
    public int _top, _left;
    public GraphicsContext _gc;
    private List<PainterAction> _actions = new ArrayList<>();

    public Painter() {
    }
    public void clear() {
        _actions.clear();
    }


    private void addAction(PainterAction action) {
        _actions.add(action);
    }

    public void setGc(GraphicsContext gc) {
        this._gc = gc;
    }

    public void fillRect(Rect rect) {
        fillRect(rect.getLeft(), rect.getTop(), rect.getWidth(), rect.getHeight());
    }

    public void fillRect(int left, int top, int width, int height) {
        addAction(new FillRectAction(left, top, width, height));
    }
    public void setFillLinearGradient(Color color1, Color color2) {
        addAction(new FillLinearGradient(color1, color2));
    }

    public void setDrawText(int x, int y, String text) {
        addAction(new PaintTextAction(x,y,text));
    }

    public void setFill(Paint background) {
        _gc.setFill(background);
    }

    public void runActions() {
        for (PainterAction action : _actions) {
            action.paint(this);
        }
    }

    public void setTextAlign(TextAlignment align) {
        addAction(new TextAlignAction(align));

    }

    public void setPos(int left, int top) {
        this._left = left;
        this._top = top;
    }
}

File: src/main/java/sample/engine/canvasui/painter/Painter.kt
package sample.engine.canvasui.painter

import javafx.geometry.VPos
import javafx.scene.canvas.GraphicsContext
import javafx.scene.paint.*
import javafx.scene.text.TextAlignment
import sample.engine.canvasui.painter.actions.*
import sample.engine.images.Rect
import java.util.ArrayList

class Painter {
    var _top: Int = 0
    var _left: Int = 0
    var _gc: GraphicsContext? = null

    fun setGc(gc: GraphicsContext) {
        this._gc = gc
    }

    fun fillRect(rect: Rect) {

        fillRect(_left+rect.left, _top+rect.top, rect.width, rect.height)
    }

    fun fillRect(left: Int, top: Int, width: Int, height: Int) {
        _gc!!.fillRect(_left+left.toDouble(), _top+ top.toDouble(), width.toDouble(), height.toDouble())
    }

    fun setFillLinearGradient(color1: Color, color2: Color) {
        _gc!!.fill = drawVerticalLinearGradient(color1, color2)
    }

    fun setDrawText(x: Int, y: Int, text: String) {
        _gc!!.fillText(text, (_left + x).toDouble(), (_top + y).toDouble())
    }

    fun setFill(background: Paint) {
        _gc!!.fill = background
    }


    fun setTextAlign(align: TextAlignment) {
        _gc!!.textAlign = align
        _gc!!.textBaseline = VPos.CENTER

    }

    fun setPos(left: Int, top: Int) {
        this._left = left
        this._top = top
    }


    private fun drawVerticalLinearGradient(firstColor: Color, secondColor: Color): LinearGradient {
        return LinearGradient(0.0, 0.0, 0.0, 1.0, true,
                CycleMethod.REFLECT,
                Stop(0.0, firstColor),
                Stop(1.0, secondColor))
    }
}